File: typescript/packages/core/src/commands/builtin/box/cut.ts
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

import type { BoxAccessor } from '../../../accessor/box.ts'
import { resolveGlob } from '../../../core/box/glob.ts'
import { read as boxRead } from '../../../core/box/read.ts'
import { IOResult, type ByteSource } from '../../../io/types.ts'
import { ResourceName, type PathSpec } from '../../../types.ts'
import { command, type CommandFnResult, type CommandOpts } from '../../config.ts'
import { specOf } from '../../spec/builtins.ts'
import { cutBytes, parseCutRanges } from '../cut_helper.ts'
import { readStdinAsync } from '../utils/stream.ts'

const ENC = new TextEncoder()

async function cutCommand(
  accessor: BoxAccessor,
  paths: PathSpec[],
  _texts: string[],
  opts: CommandOpts,
): Promise<CommandFnResult> {
  const fields = typeof opts.flags.f === 'string' ? parseCutRanges(opts.flags.f) : null
  const chars = typeof opts.flags.c === 'string' ? parseCutRanges(opts.flags.c) : null
  const delim = typeof opts.flags.d === 'string' ? opts.flags.d : '\t'
  const complement = opts.flags.complement === true
  const zero = opts.flags.z === true

  let raw: Uint8Array | null = null
  if (paths.length > 0) {
    const resolved = await resolveGlob(accessor, paths, opts.index ?? undefined)
    const first = resolved[0]
    if (first === undefined) return [null, new IOResult()]
    raw = await boxRead(accessor, first, opts.index ?? undefined)
  } else {
    raw = await readStdinAsync(opts.stdin)
    if (raw === null) {
      return [null, new IOResult({ exitCode: 1, stderr: ENC.encode('cut: missing operand\n') })]
    }
  }
  const out: ByteSource = cutBytes(raw, delim, fields, chars, complement, zero)
  return [out, new IOResult()]
}

export const BOX_CUT = command({
  name: 'cut',
  resource: ResourceName.BOX,
  spec: specOf('cut'),
  fn: cutCommand,
})

File: python/mirage/commands/builtin/databricks_volume/cut.py
# ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
# ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

from collections.abc import AsyncIterator

from mirage.accessor.databricks_volume import DatabricksVolumeAccessor
from mirage.cache.index import IndexCacheStore
from mirage.commands.builtin.databricks_volume._helpers import (
    path_prefix, read_stream_with_index)
from mirage.commands.builtin.generic.cut import cut as generic_cut
from mirage.commands.registry import command
from mirage.commands.spec import SPECS
from mirage.core.databricks_volume.glob import resolve_glob
from mirage.io.types import ByteSource, IOResult
from mirage.types import PathSpec


@command("cut", resource="databricks_volume", spec=SPECS["cut"])
async def cut(
    accessor: DatabricksVolumeAccessor,
    paths: list[PathSpec],
    *texts: str,
    stdin: AsyncIterator[bytes] | bytes | None = None,
    f: str | None = None,
    d: str | None = None,
    c: str | None = None,
    complement: bool = False,
    z: bool = False,
    index: IndexCacheStore = None,
    **_extra: object,
) -> tuple[ByteSource | None, IOResult]:
    if paths:
        paths = await resolve_glob(accessor, paths, index)
    read_stream = read_stream_with_index(index, path_prefix(paths))
    return await generic_cut(paths,
                             read_stream=read_stream,
                             accessor=accessor,
                             stdin=stdin,
                             f=f,
                             d=d,
                             c=c,
                             complement=complement,
                             z=z)
